python/private/pypi/uv_toolchain.bzl

```diff
@@ -0,0 +1,66 @@
+# Copyright 2024 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This module implements the uv toolchain rule"""
+
+UvToolchainInfo = provider(
+    doc = "Information about how to invoke the uv executable.",
+    fields = {
+        "binary": "uv binary",
+        "version": "uv version",
+    },
+)
+
+def _uv_toolchain_impl(ctx):
+    binary = ctx.executable.uv
+
+    templatevariableinfo = platform_common.TemplateVariableInfo({
+        "UV_BIN": binary.path,
+    })
+    defaultinfo = DefaultInfo(
+        files = depset([binary]),
+        runfiles = ctx.runfiles(files = [binary]),
+    )
+    uvtoolchaininfo = UvToolchainInfo(
+        binary = binary,
+        version = ctx.attr.version.removeprefix("v"),
+    )
+
+    # Export all the providers inside our ToolchainInfo
+    # so the resolved_toolchain rule can grab and re-export them.
+    toolchaininfo = platform_common.ToolchainInfo(
+        uvtoolchaininfo = uvtoolchaininfo,
+        templatevariableinfo = templatevariableinfo,
+        defaultinfo = defaultinfo,
+    )
+    return [
+        defaultinfo,
```

Comment:
> I guess it lets you put a specific toolchain implementation as a data dependency? That seems super niche, though. It would have to be a case where you didn't want to go through toolchain resolution, you knew exactly what concrete toolchain you wanted to reference, and you wanted to use its default outputs for some reason.

Pretty much this. We don't know how users will use toolchains and we may as well give them flexibility via a `data` dependency. It's a bit similar to that `rustfmt` conversation. These "prebuilt" / "hermetic" toolchains get shipped all of the place.